### PR TITLE
Map events 604-605 for checking if a specific Techno enters in a cell

### PR DESCRIPTION
**Pull request by FS-21**
Use them in Celltags (I' not a map maker so I don't if can be used in other places).

- 604: Checks if the techno that entered in the cell has the same ID specified in the event.
- 605: Checks if the techno that entered in the cell appears in the selected list in `AITargetTypes`.
- `HouseIndex` can be customized to focus in a specified house.

In `mycampaign.map`:
```ini
[Events]
...
ID=EventCount,...,604,2,[HouseIndex],[TechnoType],...
ID=EventCount,...,605,2,[HouseIndex],[AITargetTypes index#],...
...
```

| *House Index* | *Description*                              |
| :-----------: | :----------------------------------------: |
| >= 0          | The index of the current House in the map  |
| -1            | This value is ignored (any house is valid) |
| -2            | Pick the owner of the map trigger          |

### CODE REVIEW

**Review by Sovietianqi — APPROVED** [review]

File: src/Ext/TEvent/Body.cpp [hunk 1]
```diff
@@ -181,6 +185,92 @@ bool TEventExt::HouseDoesntOwnTechnoTypeTEvent(TEventClass* pThis)
 	return !TEventExt::HouseOwnsTechnoTypeTEvent(pThis);
 }
 
+bool TEventExt::CellHasAnyTechnoTypeFromListTEvent(TEventClass* pThis, ObjectClass* pObject, HouseClass* pEventHouse)
+{
+	if (!pObject)
+		return false;
```
Comment (Starkku) [inline src/Ext/TEvent/Body.cpp:191]: Is there a realistic chance that this is ever `nullptr`?

File: src/Ext/TEvent/Body.cpp [hunk 2]
```diff
@@ -181,6 +185,92 @@ bool TEventExt::HouseDoesntOwnTechnoTypeTEvent(TEventClass* pThis)
...
+	int desiredListIdx = -1;
+	if (sscanf_s(pThis->String, "%d", &desiredListIdx) <= 0 || desiredListIdx < 0)
+	{
+		Debug::Log("Error in event %d. The parameter 2 '%s' isn't a valid index value for [AITargetTypes]\n", static_cast<PhobosTriggerEvent>(pThis->EventKind), pThis->String);
+		return false;
+	}
+
+	if (RulesExt::Global()->AITargetTypesLists.size() == 0
+		|| RulesExt::Global()->AITargetTypesLists[desiredListIdx].size() == 0)
+		return false;
```
Comment (Starkku) [inline src/Ext/TEvent/Body.cpp:202]: Multiline if statements require brackets per code style guide.

File: src/Ext/TEvent/Body.cpp [hunk 3]
```diff
@@ -181,6 +185,92 @@ bool TEventExt::HouseDoesntOwnTechnoTypeTEvent(TEventClass* pThis)
...
+
+	auto const pTechno = abstract_cast<TechnoClass*>(pObject);
+	if (!pTechno)
+		return false;
```
Comment (Starkku) [inline src/Ext/TEvent/Body.cpp:206]: Add empty line before if statement.

File: src/Ext/TEvent/Body.cpp [hunk 4]
```diff
@@ -181,6 +185,92 @@ bool TEventExt::HouseDoesntOwnTechnoTypeTEvent(TEventClass* pThis)
...
+		}
+	}
+
+	return found;
+}
+
+bool TEventExt::CellHasTechnoTypeTEvent(TEventClass* pThis, ObjectClass* pObject, HouseClass* pEventHouse)
+{
+	if (!pObject)
+		return false;
```
Comment (Starkku) [inline src/Ext/TEvent/Body.cpp:236]: Same thing with the `nullptr` here.

File: src/Ext/TEvent/Body.cpp [hunk 5]
```diff
@@ -181,6 +185,92 @@ bool TEventExt::HouseDoesntOwnTechnoTypeTEvent(TEventClass* pThis)
...
+
+	auto pDesiredType = TechnoTypeClass::Find(pThis->String);
+	if (!pDesiredType)
```
Comment (Starkku) [inline src/Ext/TEvent/Body.cpp:239]: Spacing.

File: src/Ext/TEvent/Body.cpp [hunk 6]
```diff
@@ -181,6 +185,92 @@ bool TEventExt::HouseDoesntOwnTechnoTypeTEvent(TEventClass* pThis)
...
+	{
+		Debug::Log("Error in event %d. The parameter 2 '%s' isn't a valid Techno ID\n", static_cast<PhobosTriggerEvent>(pThis->EventKind), pThis->String);
+		return false;
+	}
+
+	auto const pTechno = abstract_cast<TechnoClass*>(pObject);
+	if (!pTechno)
+		return false;
```
Comment (Starkku) [inline src/Ext/TEvent/Body.cpp:247]: Spacing again.

**Review by Starkku — CHANGES_REQUESTED** [review]
Some code style nitpicks.

File: src/Ext/TEvent/Body.cpp (hunk 1 above)
Comment (FS-21) [reply src/Ext/TEvent/Body.cpp:191]: ALMOST ALWAYS, as soon a techno enters into the cell pObject gets a value so the code execution can continue.

File: src/Ext/TEvent/Body.cpp (hunk 2 above)
Comment (FS-21) [reply src/Ext/TEvent/Body.cpp:202]: I didn't notice, thnx!

File: src/Ext/TEvent/Body.cpp (hunk 4 above)
Comment (FS-21) [reply src/Ext/TEvent/Body.cpp:236]: As mentioned before this is mandatory, only gets value when an object enters into the cell.

File: src/Ext/TEvent/Body.cpp (hunk 5 above)
Comment (FS-21) [reply src/Ext/TEvent/Body.cpp:239]: Not necessary due to the code style guide, no? :-/
image

File: src/Ext/TEvent/Body.cpp (hunk 6 above)
Comment (FS-21) [reply src/Ext/TEvent/Body.cpp:247]: image

**Review by Starkku — APPROVED** [review]

**Comment by Starkku** [comment]
Fix the merge conflicts and it is good to go.

**Pull request merged**